proposals/defer-hydration.md

```diff
@@ -0,0 +1,90 @@
+# Defer Hydration Protocol
+
+An open protocol for controlling hydration on the client.
+
+Author: Justin Fagnani
+
+Status: Draft
+
+Last update: 2021-06-24
+
+# Background
+
+In server-side rendered (SSR) applications, the process of a component running code to re-associate its template with the server-rendered DOM is called "hydration". The defer-hydration protocol is design to allow controler over hydration to solve two related problems:
+
+1. Interoperable incremental hydration across web components. Componentents should not automatically hydrate upon being defined, but wait for a signal from their parent or other coordinator.
```

Comment:
Can you explain why this is (why they shouldn't automatically hydrate)? Is there an example where doing so would create unwanted results?

Reply:
I could see this fitting into a "when work gets done" topic of conversation by allowing islands to be linked generally by component name, while still being able to disambiguate on which of those islands are _needed_ at any one time. Much like Astrobuild leverages the `:visible` modifier to control when the code for a component is loaded, this could allow other instances of the component on the same page the ability upgrade with the initially visible instance, and then wait to hydrate until they, too, are in view.

I think this sort of laziness deserves to be revisited more thoroughly a the browser spec level. This seems roughly in line with the a proposal around lazy definitions: https://github.com/WICG/webcomponents/issues/782 and could arguably fit into a larger conversation around the browser allowing for lazy upgrades, which would be somewhat similar to what this protocol looks to support.

Regardless, if you're SSR'ing the entirety of an element that _may_ not come into view, the ability to control when it reconciles that delivered SSR render with the client-side render sounds like a good idea.

Reply:
Sorry for just now getting back to this, but why would you want to delay rendering a component? The reason for those modifiers are to:

1. Deprioritize loading of the code so that more important assets can be loaded over network.
2. Deprioritize executing the component (and its dependencies) JS so that more important assets can use the CPU instead.

The reason is not to delay *rendering*. Rendering should be relatively cheap, especially if the component was prerendered, there's no work to do but set up the event listeners.

It is, of course, possible that some components are going to be more expensive than others, but that's reason for those particular components to have their own API to delay rendering, not to make it a standard that is used arbitrarily.

The goal is to make the page as interactive as possible. Delaying rendering without good reason should be avoided.

Reply:
Again, the main reason is to enable top-down initialization, even for spontaneous first renders. We need our components to initialize in the same order whether CSR'ed or SSR'ed, independent of definition order.

Reply:
That's a different use case (being discussed below). The island orchestration use case is another, it's why Eleventy/webc has adopted this for example.